# Literature-mining
## The code here can be used to reproduce the results of bioinformatic analysis from this original publishing: https://iubmb.onlinelibrary.wiley.com/doi/10.1002/biof.1774
### Retrieving possible O-GlcNAcylated proteins that correlated with Runx2 expression and the corresponding functional enrichment analysis
Linear and nonlinear correlations were determined using Pearson's correlation and Spearman's correlation, respectively, among all O-GlcNAc bands, Runx2, and OGT using GraphPad Prism 8 for Windows (GraphPad Software, Inc., San Diego, CA). The principal components analysis (PCA) was applied to show the different O-GlcNAcylation patterns among DMSO, OSMI-1, and Thiamet-G groups. Since the role of Runx2 in osteoblast differentiation is well documented, the possible O-GlcNAcylated proteins with a similar molecular weight (MW) to the Runx2-correlated O-GlcNAc bands were used for a bioinformatic analysis to screen key candidate genes or signaling pathways for O-GlcNAcylation-mediated osteoblast differentiation.

Protein network data (scored links between proteins) was downloaded from the Search Tool for the Retrieval of Interacting Genes database (STRING)37 that provides comprehensive information on the functional interactions among quired proteins. Using a minimum required interaction score of 0.500 (relative higher confidence), the proteins directly associated with Runx2 were retrieved (219 proteins). Then the proteins that its predicted MW within the MW range of Runx2-correlated O-GlcNAc bands were selected (124 proteins) to check that whether or not these proteins could be O-GlcNAcylated using O-GlcNAcAtlas (https://oglcnac.org/). O-GlcNAcAtlas is the first comprehensive, and curated database encapsulating experimentally identified O-GlcNAc sites and proteins in the past 35 years (since its discovery in the early 1980s).

After searching the O-GlcNAcAtlas database, 34 proteins have recorded the O-GlcNAc site. A literature mining of the PubMed database was applied to these 34 proteins to extract the proteins that highly correlated with osteoblast differentiation. A variable R was imported to evaluate the correlation between the retrieved possible O-GlcNAcylated proteins and osteoblast differentiation by calculating the ratio of (osteoblast differentiation and quired protein-related articles)/(quired protein-related articles). We further applied p-value to statistically analyze the chance probability of co-occurrences of each protein and OD in at least k papers, as following:

![formula](https://render.githubusercontent.com/render/math?math=\mathit%20P%20=%201-\sum%\limits^{k-1}_{i=0}%20f(i)=1-\sum\limits ^{k-1}_{i=0}\frac{\binom{k}{i}\binom{N-K}{n-i}}{\binom{N}{n}})

where N is the total number of articles in PubMed (N = 32,511,529, accessed on May 06, 2021), K stands for the number of literature related osteoblast differentiation (K = 11,748), n and k represent the amounts of articles for a specific protein and its corresponding studies on OD, respectively. Subsequently, the nominal p-value was corrected as FDR using the Benjamini-Hochberg approach. In this study, proteins with FDR lower than 0.05 and R values greater than 10*K/N were selected (21 proteins) for further analysis.
Using a minimum required interaction score of 0.500 (relative higher confidence), protein–protein interaction (PPI) networks were constructed using STRING. To understand the biological significance of the retrieved proteins, the Gene Ontology (GO) and Kyoto Encyclopedia of Genes and Genomes (KEGG) pathway enrichment analysis for the 21 retrieved proteins (exclude Runx2) were performed using clusterProfiler, an R Package for Comparing Biological Themes Among Gene Clusters.41 The enriched terms with the false discovery rate (FDR) lower than 0.05 were considered as significant. The significantly enriched GO terms were further filtered using overlap coefficient (OC).42 Briefly, if the OC value between two GO terms was larger than the threshold, the GO terms with smaller FDR were kept. All related GO terms and UniProt Keywords to each retrieved protein were also collected using the DAVID and UniProt database. A word frequency analysis for the GO terms and UniProt Keywords was performed based on a Word-Cloud plot.

GO (http://www.geneontology.org) is a tool used for the unification of biological functions based on gene annotation information, which primarily consists of biological process (BP), molecular function (MF), and cellular component (CC) analyses. The KEGG (available at http://www.genome.ad.jp/kegg/) is a pathway-associated database connecting known information on molecular interaction networks. The term, “Keywords,” in UniPort summarizes the content of a UniProtKB entry and facilitates the search for proteins of interest.
